File: 0958-check-completeness-of-a-binary-tree/0958-check-completeness-of-a-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int totalNodes;
    int countNodes(TreeNode* root){
        if(root == NULL) return 0;
        return 1 + countNodes(root->left) + countNodes(root->right);
    }
    bool helper(TreeNode* root, int idx){
        if(root == NULL) return true;
        if(idx > totalNodes) return false;
        return helper(root->left, 2 * idx) && helper(root->right, 2 * idx + 1);
    }
    bool isCompleteTree(TreeNode* root) {
        totalNodes = countNodes(root);
        return helper(root, 1);
    }
};

File: 0958-check-completeness-of-a-binary-tree/0958-check-completeness-of-a-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean isCompleteTree(TreeNode root) {
        Queue<TreeNode> q = new LinkedList<>();
        q.add(root);
        boolean flag = false;
        while(!q.isEmpty()){
            int size = q.size();            
            for(int i = 0; i < size; i++){
                TreeNode curr = q.remove();
                if(flag && (curr.left != null || curr.right != null)) return false;
                if(curr.left != null) q.add(curr.left);
                if(curr.right != null) q.add(curr.right);
                if(curr.left == null || curr.right == null) flag = true;
                if(flag && curr.left == null && curr.right != null) return false;
            }
        }
        return true;
    }
}